textdb/textdb-dataflow/src/main/java/edu/uci/ics/textdb/dataflow/regexmatch/TrigramBooleanQuery.java

```diff
@@ -0,0 +1,172 @@
+package edu.uci.ics.textdb.dataflow.regexmatch;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.StringJoiner;
+
+import edu.uci.ics.textdb.common.constants.DataConstants;
+
+/**
+ * 
+ * Trigram Query of OR and AND
+ * 
+ * {@code operandList} is a list of literals (strings) in this query.
+ * {@code subQueryList} is a list of parenthesized RegexTrigramQuery.
+ * {@code operator} is the operator connecting each literals in {@code operandList} and each subqueries in {@code subqueryList};
+ * For example, RegexTrigramQuery for regex "data(abc|bcd)" is "dat AND ata AND (abc OR bcd)"
+ * The operand of this query is AND
+ * operands = ["dat", "ata"]
+ * subQueries = ["abc OR bcd"]
+ * 
+ * The trigram query of a regex has two high-level layers:
+ * First, conjunction of TrigramBooleanQuery of prefix, suffix, and exact.
+ * Second, disjunction of TrigramBooleanQuery of each element respectively in prefix, suffix, and exact.
+ * 
+ * @Author Zuozhi Wang
+ * @Author Shuying Lai
+ * 
+ */
+public class TrigramBooleanQuery {
+	public static final int NONE = 0;
+	public static final int ANY  = 1;
+	public static final int AND  = 2;
+	public static final int OR   = 3;
+	
+	/**
+	 * operator is NONE/ALL/AND/OR
+	 */
+	int operator;
+	List<String> operandList;
```

Comment:
One thing we need to do with GramBooleanQuery is remove duplicated "leaves" to simplify the query. In this case we need to quickly find out all atomic operands.  If a `hashCode()` design could fulfill this requirement, I think it ok to take the second design.
Although we haven't figure out the necessity of using `operandList` and `subQueriesList`, in general, it will be convenient for us if we could follow the data structures of Russ Cox's implementation (easier to implement and debug),